datafusion/core/src/physical_optimizer/pruning.rs

```diff
@@ -478,6 +478,31 @@ pub struct PruningPredicate {
     literal_guarantees: Vec<LiteralGuarantee>,
 }
 
+/// Hook to handle predicates that DataFusion can not handle, e.g. certain complex expressions
+/// or predicates that reference columns that are not in the schema.
+pub trait UnhandledPredicateHook {
+    /// Called when a predicate can not be handled by DataFusion's transformation rules
+    /// or is referencing a column that is not in the schema.
+    fn handle(&self, expr: &Arc<dyn PhysicalExpr>) -> Arc<dyn PhysicalExpr>;
+}
```

Comment:
This could be a closure but I had issues with lifetimes, etc. Having the trait also gives it a useful name 😄 

The other API questions are:
- Should this be mutable? I think implementers can just use interior mutability if needed.
- Should this make it easier to say "use the existing expression"? I don't think that's a common case, and the current APIs use `&Arc<dyn PhysicalExpr> -> Arc<dyn PhysicalExpr>` as well. Plus it's as easy as a Clone on an Arc.